jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -182,41 +182,39 @@ public static URL computePURootURL(URL pxmlURL, String descriptorLocation) throw
             // e.g. file:/tmp/META-INF/persistence.xml
             // 210280: any file url will be assumed to always reference a file (not a directory)
             result = new URL(pxmlURL, path.toString()); // NOI18N
-        } else if("jar".equals(protocol)) { // NOI18N
-            // e.g. jar:file:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            JarURLConnection conn =
-                    JarURLConnection.class.cast(pxmlURL.openConnection());
-            result = conn.getJarFileURL();
-        } else if("zip".equals(protocol)) { // NOI18N
-            // e.g. zip:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            // stolen from java.net.JarURLConnection.parseSpecs method
-            String spec = pxmlURL.getFile();
-            int separator = spec.lastIndexOf("!/");
-            if (separator == -1) {
-                separator = spec.length() - 1;
-            }
-            result = new File(spec.substring(0, separator++)).toURL();
-        } else if("wsjar".equals(protocol)) { // NOI18N
-            // e.g. wsjar:file:/tmp/a_ear/b.jar!/META-INF/persistence.xml
-            // but WS gives use jar:file:..., so we need to match it.
-            String spec = pxmlURL.getFile();
+        } else if("zip".equals(protocol) ||
+                  "jar".equals(protocol) ||
+                  "wsjar".equals(protocol)) {
+            // Example URLs and expected results:
+            // zip:/foo/bar.jar!/META-INF/persistence.xml -> file:/foo/bar.jar
+            // zip:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml -> jar:file:/foo/bar.jar!/WEB-INF/classes
+            // jar:file:/foo/bar.jar!/META-INF/persistence.xml -> file:/foo/bar.jar
+            // jar:file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml -> jar:file:/foo/bar.jar!/WEB-INF/classes
+            // wsjar:file:/foo/bar.jar!/META-INF/persistence.xml -> file:/foo/bar.jar
+            // wsjar:file:/foo/bar.war!/WEB-INF/classes/META-INF/persistence.xml -> wsjar:file:/foo/bar.jar!/WEB-INF/classes
```

Comment:
It would be nice if we had these as tests in PersistenceUnitProcessorTest :)

Reply:
Also negative tests for invalid scenarios like ```wsjar:file:/a/path/to/my.war!/foo/WEB-INF/classes/META-INF/persistence.xml``` curious how that is handled.

Reply:
Sure thing. Promoted all of these to test cases.